docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,206 @@
+
+The second is generated by the extension controllers in the seed and then synced to the shoot by a new ShootState controller (described below).
+
+```
+apiVersion: core.gardener.cloud/v1alpha1
+kind: ShootState
+metadata:
+  name: my-shoot
+  namespace: garden-core
+gardener:
+- name: ca
+  data:
+    ca.crt: ...
+    ca.key: ...
+- name: ssh-keypair
+  data:
+    id_rsa: ...
+- name:
+...
+extensions:
+- kind: Infrastructure
+  state: ... (Terraform state)
+- kind: ControlPlane
+  purpose: normal
+  state: ... (Certificates generated by the extension)
+- kind: Worker
+  state: ... (Machine objects)
+```
+
+### Gardener extensions changes
+
+All extension controllers which support state migration must have a new `status.state` field and support an annotation `core.gardener.cloud/operation=restore` in the respective Custom Resources.
+
+The state field is populated with whatever data the extension controller requires to restore its state on a different seed cluster. As an example the infrastructure resource must save the terraform state.
+
+**Note**: some states can be too large so it makes sense to consider adding them as configmap or secret references.
+
+```
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Infrastructure
+metadata:
+  name: infrastructure
+  namespace: shoot--foo--bar
+spec:
+  type: azure
+  region: eu-west-1
+  secretRef:
+    name: cloudprovider
+    namespace: shoot--foo--bar
+  providerConfig:
+    apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
+    kind: InfrastructureConfig
+    resourceGroup:
+      name: mygroup
+    networks:
+      vnet: # specify either 'name' or 'cidr'
+      # name: my-vnet
+        cidr: 10.250.0.0/16
+      workers: 10.250.0.0/19
+status:
+  state:
+    terraformState: |
+      {
+          "version": 3,
+          "terraform_version": "0.11.14",
+          "serial": 2,
+          "lineage": "3a1e2faa-e7b6-f5f0-5043-368dd8ea6c10",
+          "modules": [
+              {
+              }
+          ]
+          ...
+      }
+```
+
+
+### Shoot reconciliation flow changes
+
+The only data which must be stored in the `ShootState` by the Garden Controller Manager are secrets (e.g ca for the apiserver). Therefore the `botanist.DeploySecrets` step is changed. It is split into two functions which take a list of secrets that have to be generated.
+- `botanist.GenerateSecretState` Generates certificate authorities and other secrets which have to be persisted in the ShootState and must not be regenerated on the New Seed.
+- `botanist.DeploySecrets` Takes secret data from the `ShootState`, generates new ones (e.g. client tls certificates from the saved certificate authorities) and deploys everything in the Shoot's control plane on the New Seed
+
```

Comment:
I like the split, one part writes the state (in this case, rather some desired state), the other generates whatever is needed out of that (the technical artefacts). That however, is not always possible, like with Terraform where the TF file is generated from other data and the state only gets created after running TF. Either way, two means to the same end, the state needed to recreate everything is in the garden cluster and that's good.